server/vulnerabilities/cve.go

```diff
@@ -148,9 +181,38 @@ func checkCVEs(ctx context.Context, ds fleet.Datastore, logger kitlog.Logger, cp
 							}
 							matchingCPEs = append(matchingCPEs, cpe)
 						}
-						err = ds.InsertCVEForCPE(ctx, matches.CVE.ID(), matchingCPEs)
+
+						newCount, err := ds.InsertCVEForCPE(ctx, matches.CVE.ID(), matchingCPEs)
 						if err != nil {
 							level.Error(logger).Log("cpe processing", "error", "err", err)
+							continue // do not report a recent vuln that failed to be inserted in the DB
+						}
+
+						// collect as recent vuln only if newCount > 0, otherwise we would send
+						// webhook requests for the same vulnerability over and over again until
+						// it is older than 2 days.
+						if collectVulns && newCount > 0 {
+							vuln, ok := matches.CVE.(*feednvd.Vuln)
+							if !ok {
+								level.Error(logger).Log("recent vuln", "unexpected type for Vuln interface", "type", fmt.Sprintf("%T", matches.CVE))
+								continue
+							}
+
+							if rawPubDate := vuln.Schema().PublishedDate; rawPubDate != "" {
+								pubDate, err := time.Parse(publishedDateFmt, rawPubDate)
+								if err != nil {
+									level.Error(logger).Log("recent vuln", "unexpected published date format", "published_date", rawPubDate, "err", err)
+									continue
+								}
+								// the second condition should only affect tests - to ignore pubDates in the future
+								// when using a mocked current clock. When using the real clock, the published date
+								// should always be in the past.
+								if theClock.Since(pubDate) <= recentVulnMaxAge && theClock.Now().After(pubDate) {
+									mu.Lock()
+									recentVulns[matches.CVE.ID()] = matchingCPEs
```

Comment:
```go
recentVulns[matches.CVE.ID()] = append(recentVulns[matches.CVE.ID()], matchingCPEs...)
```
Though, not sure if really needed how things work here (A CVE is only processed once?)

Reply:
Actually I think it's needed, this goroutine processes by CPEs.

Reply:
Good catch!